modules/internal/ChapelArray.chpl

```diff
@@ -2202,7 +2196,7 @@ module ChapelArray {
      */
     proc safeCast(type t:_domain)
       where chpl__isRectangularDomType(t) && isRectangularDom(this) {
-      const tmpD: t;
+      var tmpD: t;
       if tmpD.rank != this.rank then
```

Comment:
Casts produce RHS things. In other words, the domain returned by the cast is `const`.
So let us add @e-kayrakli's `chpl__fixupConstDomain(tmpD, true)` before returning `tmpD`

Perhaps obviously, chpl__fixupConstDomain should not be applied to `this` when it is returned.

Reply:
> Casts produce RHS things. In other words, the domain returned by the cast is `const`.

Ala C++11 Rvalues?

> So let us add @e-kayrakli's `chpl__fixupConstDomain(tmpD, true)` before returning `tmpD`

Does Chapel have the equivalent of C++ [NRVO](https://en.cppreference.com/w/cpp/language/copy_elision) (Named Returned Value Optimization), where  a local variable is "moved" in a `return` instead of copied? That seems a more complete solution than `const`.

> Perhaps obviously, chpl__fixupConstDomain should not be applied to this when it is returned.

Unless Chapel has `this` intent equivalent to Rvalue reference-qualified members, e.g.,
```c++
class C {
public:
    C func() &&   // <-- Rvalue-qualified *this
    {
      // ...
      return std::move(*this);  // Moves *this instead of copying it
    }
};
```

Reply:
@leekillough - first, I apologize for giving you extra work. Turns out - see below - that `chpl__fixupConstDomain(tmpD);` should not be here - in both _cast and safeCast.

Could you please take it out in both casts? Thank you!

The reason is that the result of a domain cast CAN be variable. So the class in tmpD._value can be resized. Consider this code:
```chpl
const d = {1..3};
var d2 = d: domain(1, stridable=true);   // cast here
var a: [d2] int;
d2 = {1..1_000_000};
```

In this case, d2._value is the `tmpD` from the _cast function. If we do `chpl__fixupConstDomain(tmpD)` within the _cast, then assigning to `d2` will not adjust the memory for the elements of `a`, and we will have incorrect memory accesses.

I will write separately regarding C++.

Reply:
> Does Chapel have the equivalent of C++ NRVO (Named Returned Value Optimization), where a local variable is "moved" in a return instead of copied?

When a function returns by value something that was provably created locally to the function, ex. a record or a domain or an array, the ownership to that thing is passed to the caller.

> Perhaps obviously, chpl__fixupConstDomain should not be applied to `this` when it is returned.

Here is my reasoning for this. In these cast functions, `this` intent is `const ref`. So the caller can have other references to the `domain` being passed in, and it can be non-const in the caller's context. Treating is as `const` could lead to invalid memory accesses similarly to my example above.

> Ala C++11 Rvalues?

Perhaps, I am not sure.

Reply:
> @leekillough - first, I apologize for giving you extra work. Turns out - see below - that `chpl__fixupConstDomain(tmpD);` should not be here - in both _cast and safeCast.
> 
> Could you please take it out in both casts? Thank you!
> 
> The reason is that the result of a domain cast CAN be variable. So the class in tmpD._value can be resized. Consider this code:
> 
> ```chapel
> const d = {1..3};
> var d2 = d: domain(1, stridable=true);   // cast here
> var a: [d2] int;
> d2 = {1..1_000_000};
> ```
> 
> In this case, d2._value is the `tmpD` from the _cast function. If we do `chpl__fixupConstDomain(tmpD)` within the _cast, then assigning to `d2` will not adjust the memory for the elements of `a`, and we will have incorrect memory accesses.

I don't understand. Does this mean the original logic of `chpl__fixupConstDomain()` is not valid?

I thought `d2` was the domain, and `a` was indexed by `d2`, so you can change `d2` without affecting `a`. When `a` is created, it takes a copy of `d2` as its domain.

Cc: @e-kayrakli

Reply:
> Does this mean the original logic of chpl__fixupConstDomain() is not valid?

No, this has no impact on chpl__fixupConstDomain() itself.

> you can change `d2` without affecting `a`. When a is created, it takes a copy of `d2` as its domain

It is exactly the opposite. Chapel's design is that the set of indices of an array is controlled by the domain over which it is declared. The assignment to `d2` at the end of my code snippet, in addition to changing d2's value, also results in resizing `a` from three to million elements.

The point of chpl__fixupConstDomain() when it is passed `true` is to indicate that the argument domain will not be resized ever. This eliminates some overhead in the domain implementation.

The point of my belated realization is that `tmpD` within the cast functions CAN be resized, ex. in my code snippet. Therefore it would be a bug if we invoked chpl__fixupConstDomain(true) on it.

Reply:
Please look at the latest changes. I think we should have a meeting (not team-wide) to discuss.

Reply:
@vasslitvinov -- I have doubts about removing those calls from casts. In your example, I would like the domain on the RHS of `var d2 = d:domain(1, stridable=true)`, to be a constant domain. But when it is assigned to a `var`, it's constness is adjusted by the compiler or by virtue of `initCopy` et al.

In other words, I don't want any array tracking to happen for:

```chapel
var a: [d:domain(1, stridable=true)];
```

And without doing anything special for casts, that's not possible, I think.

Offline, Lee said the paratest is passing without those, so we can merge this PR in, and I can check whether there's a change we need to make after. At least we can add some testing for this case.

(Sorry for not seeing this conversation earlier)

Reply:
@e-kayrakli I understand your motivation. However, I disagree that we should merge code with a known bug.

Reply:
I understand.

I just tested that:

```chapel
var d = {1..10};

var a: [d:domain(1, stridable=true)] int;

writeln(a.domain.definedConst);
```

doesn't compile on master. But on this branch it compiles, only prints out false. As far as I know this just implies unnecessary performance overhead and not correctness issue. (So, the PR is still a step forwards).  But if it is straightforward to add those calls back, I think that'd be best.

Reply:
@e-kayrakli try to implement your suggestion - take this branch and add `chpl__fixupConstDomain(tmpD, true)` within the _cast() - then compile+run this code:

```chpl
const d = {1..3};
var d2 = d: domain(1, stridable=true);   // cast here
var a: [d2] int;
d2 = {1..1_000_000};
writeln(a);
```

In my experiment, it crashed even without valgrind/asan.
This makes me object adding those calls to the cast functions.